File: Blind-75/SlidingWindow/BestTimeToBuyAndSellStock/pb-121.java
class Solution {
    // O(N)
    public int maxProfit(int[] prices) {
        int res = 0;
        int l = 0;
        for (int r = 0; r < prices.length; r++){
            while(prices[r] - prices[l] < 0) l++;
            res = Math.max(res, prices[r] - prices[l]);
        }
        return res;
    }
}

File: Blind-75/SlidingWindow/BestTimeToBuyAndSellStock/pb-121.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy, sell = 0, 1 
        maxP = 0
        while sell < len(prices):
            if prices[buy] < prices[sell]: # profitable
                profit = prices[sell] - prices[buy]
                maxP = max(maxP, profit)
            else: # not profitable
                buy = sell
            sell += 1
        return maxP